torchtune/modules/tokenizer.py

```diff
@@ -89,3 +91,47 @@ def decode(self, ids: List[int]) -> str:
             str: The decoded text.
         """
         return self.spm_model.decode(ids)
+
+    def tokenize_messages(
+        self, messages: List[Message], max_seq_len: Optional[int] = None
+    ) -> Tuple[List[int], List[bool]]:
+        start_of_turn = True
+        end_of_turn = False
+        tokenized_messages = []
+        mask = []
+        for message in messages:
+            # If assistant message, this is the end of a turn
+            end_of_turn = message.role == "assistant"
+
+            # Prepend BOS on start of new turns
+            if start_of_turn:
+                tokenized_messages.append(self.bos_id)
+                mask.append(message.masked)
+                start_of_turn = False
+
+            # Tokenize current message, append with masks
+            tokens = self.spm_model.encode(
+                message.content,
+                add_bos=False,
+                add_eos=False,
+            )
+            tokenized_messages.extend(tokens)
+            mask.extend([message.masked] * len(tokens))
+
+            # Break out early if we reach max_seq_len
+            if max_seq_len and len(tokens) >= max_seq_len:
+                break
+
+            # If assistant message, append EOS at end
+            if end_of_turn:
```

Comment:
You should probably check for max seq len after the EOS id is added in case this puts it right at max seq len

Reply:
Personally I think it's not worth the extra code to check in the case we hit max_seq_len exactly given (a) it's a pretty rare edge case, (b) it adds complexity to the code, and (c) we would still break out on the next message anyways (and we still truncate everything to the right length at the end regardless). Lmk if this makes sense to you

Reply:
yeah I meant just move the if statement above to below this end_of_turn section, shouldn't be extra code. but yeah either way the result is the same, just a minor speed up. your call

Reply:
Oh duh lol. Yeah I can do that